File: llm-harness-js/src/stats/llm/matcher/pipeline.ts
import {
  FALLBACK_PROVIDER_IDS,
  PRIMARY_PROVIDER_ID,
  asRecord,
  modelSlugFromModelId,
  normalizeModelToken,
} from "../shared.js";

import {
  compareCandidates,
  hasFirstTokenMatch,
  scoreCandidate,
} from "./scoring.js";
import { splitBaseModelId } from "./tokenize.js";
import {
  type ArtificialAnalysisModel,
  type LlmMatchCandidate,
  type LlmMatchResult,
  type MatcherInputModel,
  type MatcherRunOutput,
  type ModelsDevModel,
  type PreferredProviderScopedModels,
} from "./types.js";

const VOID_THRESHOLD_RANGE_RATIO = 0.35;

export function uniqueModelCount(modelsDevModels: ModelsDevModel[]): number {
  return new Set(
    modelsDevModels.map((modelsDevModel) => modelsDevModel.model_id),
  ).size;
}

function hasExactSlugFallbackCandidate(
  artificialAnalysisSlug: string,
  fallbackCandidates: LlmMatchCandidate[],
): boolean {
  const normalizedArtificialAnalysisSlug = normalizeModelToken(
    artificialAnalysisSlug,
  );
  if (normalizedArtificialAnalysisSlug.length === 0) {
    return false;
  }
  return fallbackCandidates.some((candidate) => {
    const candidateSlug = normalizeModelToken(
      splitBaseModelId(candidate.model_id),
    );
    return (
      candidateSlug.length > 0 &&
      candidateSlug === normalizedArtificialAnalysisSlug
    );
  });
}

export function splitPreferredProviderModels(
  modelsDevModels: ModelsDevModel[],
): PreferredProviderScopedModels {
  const primary = modelsDevModels.filter(
    (modelsDevModel) => modelsDevModel.provider_id === PRIMARY_PROVIDER_ID,
  );
  const fallback = modelsDevModels.filter((modelsDevModel) =>
    FALLBACK_PROVIDER_IDS.has(modelsDevModel.provider_id),
  );
  return { primary, fallback };
}

function collectCandidatesForArtificialAnalysisSlug(
  artificialAnalysisSlug: string,
  modelsDevModels: ModelsDevModel[],
): LlmMatchCandidate[] {
  if (!artificialAnalysisSlug) {
    return [];
  }

  return modelsDevModels
    .map((modelsDevModel) => {
      const modelsDevModelName =
        typeof modelsDevModel.model.name === "string"
          ? modelsDevModel.model.name
          : "";
      if (
        !hasFirstTokenMatch(
          artificialAnalysisSlug,
          modelsDevModel.model_id,
          modelsDevModelName,
        )
      ) {
        return null;
      }
      const candidateScore = scoreCandidate(
        artificialAnalysisSlug,
        modelsDevModel.model_id,
        modelsDevModelName,
      );
      if (candidateScore <= 0) {
        return null;
      }
      return {
        model_id: modelsDevModel.model_id,
        provider_id: modelsDevModel.provider_id,
        provider_name: modelsDevModel.provider_name,
        model_name: modelsDevModelName || null,
        score: candidateScore,
      };
    })
    .filter((candidate): candidate is LlmMatchCandidate => candidate != null)
    .sort(compareCandidates);
}

function selectPreferredCandidatesForArtificialAnalysisSlug(
  artificialAnalysisSlug: string,
  scopedModels: PreferredProviderScopedModels,
): LlmMatchCandidate[] {
  const primaryCandidates = collectCandidatesForArtificialAnalysisSlug(
    artificialAnalysisSlug,
    scopedModels.primary,
  );
  const fallbackCandidates = collectCandidatesForArtificialAnalysisSlug(
    artificialAnalysisSlug,
    scopedModels.fallback,
  );
  if (primaryCandidates.length === 0) {
    return fallbackCandidates;
  }
  if (
    hasExactSlugFallbackCandidate(artificialAnalysisSlug, fallbackCandidates)
  ) {
    return fallbackCandidates;
  }
  return primaryCandidates;
}

function applyMaxMinHalfVoid<
  T extends { best_match: LlmMatchResult; candidates?: unknown[] },
>(models: T[]): { threshold: number | null; voided: number } {
  const scores = models
    .map((model) => model.best_match?.score)
    .filter((score): score is number => Number.isFinite(score))
    .sort((left, right) => left - right);
  if (scores.length === 0) {
    return { threshold: null, voided: 0 };
  }

  const minScore = scores[0] as number;
  const maxScore = scores.at(-1) as number;
  const threshold =
    minScore + (maxScore - minScore) * VOID_THRESHOLD_RANGE_RATIO;
  let voided = 0;
  for (const model of models) {
    const score = model.best_match?.score;
    if (score != null && score < threshold) {
      model.best_match = null;
      if ("candidates" in model && Array.isArray(model.candidates)) {
        model.candidates = [];
      }
      voided += 1;
    }
  }
  return { threshold, voided };
}

export function runMatcher(
  sourceModels: MatcherInputModel[],
  scopedModels: PreferredProviderScopedModels,
  maxCandidates: number,
): MatcherRunOutput {
  const models = sourceModels.map((sourceModel) => {
    const candidates = selectPreferredCandidatesForArtificialAnalysisSlug(
      sourceModel.artificialAnalysisSlug,
      scopedModels,
    ).slice(0, maxCandidates);
    return {
      artificial_analysis_slug: sourceModel.artificialAnalysisSlug,
      artificial_analysis_name: sourceModel.artificialAnalysisName,
      artificial_analysis_release_date:
        sourceModel.artificialAnalysisReleaseDate,
      best_match: candidates[0] ?? null,
      candidates,
    };
  });

  const preVoidMatchedCount = models.filter(
    (model) => model.best_match != null,
  ).length;
  const preVoidUnmatchedCount = models.length - preVoidMatchedCount;
  const voidStats = applyMaxMinHalfVoid(models);
  const matchedCount = models.filter(
    (model) => model.best_match != null,
  ).length;
  const unmatchedCount = models.length - matchedCount;

  return {
    models,
    voidThreshold: voidStats.threshold,
    voidedCount: voidStats.voided,
    preVoidMatchedCount,
    preVoidUnmatchedCount,
    matchedCount,
    unmatchedCount,
  };
}

export function buildInputModelsFromArtificialAnalysis(
  artificialAnalysisModels: ArtificialAnalysisModel[],
): MatcherInputModel[] {
  return artificialAnalysisModels.map((artificialAnalysisModel) => ({
    artificialAnalysisSlug:
      typeof artificialAnalysisModel.slug === "string"
        ? artificialAnalysisModel.slug
        : "",
    artificialAnalysisName:
      typeof artificialAnalysisModel.name === "string"
        ? artificialAnalysisModel.name
        : null,
    artificialAnalysisReleaseDate:
      typeof artificialAnalysisModel.release_date === "string"
        ? artificialAnalysisModel.release_date
        : null,
  }));
}

export function buildInputModelsFromScrapedRows(
  scrapedRows: unknown[],
): MatcherInputModel[] {
  return scrapedRows.map((scrapedRow) => {
    const scrapedRowRecord = asRecord(scrapedRow);
    const modelId =
      typeof scrapedRowRecord.model_id === "string"
        ? scrapedRowRecord.model_id
        : null;
    const artificialAnalysisSlug = modelSlugFromModelId(modelId) ?? "";
    return {
      artificialAnalysisSlug,
      artificialAnalysisName: modelId,
      artificialAnalysisReleaseDate: null,
    };
  });
}

File: llm_harness/stats/llm/matcher/pipeline.py
"""Internal native Python matcher pipeline."""

from __future__ import annotations

from ..shared import FALLBACK_PROVIDER_IDS, PRIMARY_PROVIDER_ID, normalize_model_token
from .scoring import compare_candidates, has_first_token_match, score_candidate
from .tokenize import split_base_model_id

VOID_THRESHOLD_RANGE_RATIO = 0.35


def unique_model_count(models_dev_models: list[dict]) -> int:
    """Helper for unique model count."""
    return len({model.get("model_id") for model in models_dev_models})


def _has_exact_slug_fallback_candidate(source_slug: str, fallback_candidates: list[dict]) -> bool:
    """Return whether any fallback candidate matches the source slug exactly."""
    normalized_source_slug = normalize_model_token(source_slug)
    if not normalized_source_slug:
        return False
    for candidate in fallback_candidates:
        candidate_slug = normalize_model_token(split_base_model_id(candidate["model_id"]))
        if candidate_slug and candidate_slug == normalized_source_slug:
            return True
    return False


def split_preferred_provider_models(models_dev_models: list[dict]) -> dict:
    """Split the preferred provider models."""
    return {
        "primary": [model for model in models_dev_models if model.get("provider_id") == PRIMARY_PROVIDER_ID],
        "fallback": [model for model in models_dev_models if model.get("provider_id") in FALLBACK_PROVIDER_IDS],
    }


def _collect_candidates_for_source_slug(source_slug: str, models_dev_models: list[dict]) -> list[dict]:
    """Helper for collect candidates for source slug."""
    if not source_slug:
        return []
    candidates = []
    for models_dev_model in models_dev_models:
        model_name = models_dev_model.get("model", {}).get("name") if isinstance(models_dev_model.get("model"), dict) else ""
        model_name = model_name if isinstance(model_name, str) else ""
        if not has_first_token_match(source_slug, models_dev_model.get("model_id", ""), model_name):
            continue
        candidate_score = score_candidate(source_slug, models_dev_model.get("model_id", ""), model_name)
        if candidate_score <= 0:
            continue
        candidates.append(
            {
                "model_id": models_dev_model.get("model_id", ""),
                "provider_id": models_dev_model.get("provider_id", ""),
                "provider_name": models_dev_model.get("provider_name", ""),
                "model_name": model_name or None,
                "score": candidate_score,
            }
        )
    return sorted(candidates, key=compare_candidates)


def _select_preferred_candidates_for_slug(source_slug: str, provider_pools: dict) -> list[dict]:
    """Select the preferred candidates for slug."""
    primary_candidates = _collect_candidates_for_source_slug(source_slug, provider_pools["primary"])
    fallback_candidates = _collect_candidates_for_source_slug(source_slug, provider_pools["fallback"])
    if not primary_candidates:
        return fallback_candidates
    if _has_exact_slug_fallback_candidate(source_slug, fallback_candidates):
        return fallback_candidates
    return primary_candidates


def _apply_maxmin_half_void(models: list[dict]) -> tuple[float | None, int]:
    """Apply the maxmin half void."""
    scores = sorted(model["best_match"]["score"] for model in models if model.get("best_match") is not None)
    if not scores:
        return None, 0
    min_score = scores[0]
    max_score = scores[-1]
    threshold = min_score + (max_score - min_score) * VOID_THRESHOLD_RANGE_RATIO
    voided = 0
    for model in models:
        score = model["best_match"]["score"] if model.get("best_match") is not None else None
        if score is not None and score < threshold:
            model["best_match"] = None
            if isinstance(model.get("candidates"), list):
                model["candidates"] = []
            voided += 1
    return threshold, voided


def run_matcher(source_models: list[dict], provider_pools: dict, max_candidates: int) -> dict:
    """Helper for run matcher."""
    models = []
    for source_model in source_models:
        candidates = _select_preferred_candidates_for_slug(source_model["source_slug"], provider_pools)[:max_candidates]
        models.append(
            {
                "artificial_analysis_slug": source_model["source_slug"],
                "artificial_analysis_name": source_model["source_name"],
                "artificial_analysis_release_date": source_model["source_release_date"],
                "best_match": candidates[0] if candidates else None,
                "candidates": candidates,
            }
        )
    pre_void_matched_count = len([model for model in models if model["best_match"] is not None])
    pre_void_unmatched_count = len(models) - pre_void_matched_count
    void_threshold, voided_count = _apply_maxmin_half_void(models)
    matched_count = len([model for model in models if model["best_match"] is not None])
    unmatched_count = len(models) - matched_count
    return {
        "models": models,
        "void_threshold": void_threshold,
        "voided_count": voided_count,
        "pre_void_matched_count": pre_void_matched_count,
        "pre_void_unmatched_count": pre_void_unmatched_count,
        "matched_count": matched_count,
        "unmatched_count": unmatched_count,
    }
